Readme.md
->make command in the terminal creates the a.out executable file for running the shell
-> main.c:
  . Shell implementation
  . Signal handler 

-> headers.h:
  . All the headers required are inserted here it self

-> functions.h
  . All the global variables and functions are ideclared here

-> functions.c:
  . string tokeninizig function
  . cd , pinfo, echo  implementations
  . setenv,unsetenv,jobs,kjob has been implemneted
  .fg and bg , overkill, quit

-> ls.c:
  . whole ls implementation has been done here

-> history.c:
  . History recording and history command implementation has been coded here

->History command implementation is acheived through creating a txt file where the executable file is running, it is modified every time when a command     given to  the  terminal.

-> If & is given with builtin commands like(ls &, cd &, pid &,...), they have been assumed as foreground ....

->ctrl-z,ctrl-c has been handled by creating handlers

->last working directory is achieved through storing cwd after     implentation of command.


history.txt
1
setenv 
jobs
quit
gedit 
gedit
gedit
gedit
subl
firefox
jobs
overkill
quit

quit


quit
setenv foo bar
getenv foo
setenv foo


ls.c
#include "functions.h"
#include "headers.h"

int roundNo(float num) 
{ 
    return num < 0 ? num - 0.5 : num + 0.5; 
} 


int lexi_sort(const struct dirent **a ,const struct dirent **b)
{   int p= strcasecmp( (*a)->d_name,(*b)->d_name);
    if(p>0)
     return 1;
    else if(p<0)
     return -1;
    else 
     return 0;
}

char * time_as_req(time_t tmp)
{  char *buff = malloc(100);
   struct tm * tm_tmp = localtime(&tmp);
 
strftime(buff,100,"%b %d %H:%M",tm_tmp);
return buff;
}


int check_fordot(const struct dirent *a)
{     if(strncmp(a->d_name,".",1)==0)
        return 0;
      else
       return 1;

}

char * path(char * parent_path,char *file_name)
{    char * absol_path = malloc(100*sizeof(char));
     if(file_name[0]=='/')
     return file_name;
     else
    { strcpy(absol_path,parent_path);
     strcat(absol_path,"/");
     strcat(absol_path,file_name);
     return absol_path;
    }
}

void display_short(struct dirent **dir_list,int cnt,char *parent_path)
{    for (int i=0;i<cnt;i++)
     { printf("%s\n",dir_list[i]->d_name);} 
   
}


char *detecting_parent(char *cwd_name)
{   int tmp;
    //printf("%s\n",cwd_name);
    for(int i=strlen(cwd_name)-1;i>=0;i--)
     {  if(cwd_name[i]=='/')
         {tmp = i;break;}

     }
   char * parent = malloc(tmp*sizeof(char));
     strncpy(parent,cwd_name,tmp);
  return parent;
}






void display_long(struct dirent **dir_list,int cnt,char *parent_path)
{   struct stat tmp_stat;
    stat(parent_path,&tmp_stat);
    //printf("total %ld\n",tmp_stat.st_blocks/2);
    char *lines[cnt];
    float total = 0;
    
    for(int i=0;i<cnt;i++)    
    {struct dirent *cur_dir = dir_list[i];
    struct stat cur_stat;
    stat(path(parent_path,cur_dir->d_name),&cur_stat);
    char type;
    if(S_ISDIR(cur_stat.st_mode))
    type='d';
    else if(S_ISLNK(cur_stat.st_mode))
    type ='l';
    else
    type = '-';


    total+=cur_stat.st_blocks; 
    char permissions[9];
  
    permissions[0]=(cur_stat.st_mode & S_IRUSR) ? 'r' : '-';
    permissions[1]=(cur_stat.st_mode & S_IWUSR) ? 'w' : '-';
    permissions[2]=(cur_stat.st_mode & S_IXUSR) ? 'x' : '-';
    permissions[3]=(cur_stat.st_mode & S_IRGRP) ? 'r' : '-';
    permissions[4]= (cur_stat.st_mode & S_IWGRP) ? 'w' : '-';
    permissions[5]= (cur_stat.st_mode & S_IXGRP) ? 'x' : '-';
    permissions[6]=(cur_stat.st_mode & S_IROTH) ? 'r' : '-';
    permissions[7]= (cur_stat.st_mode & S_IWOTH) ? 'w' : '-';
    permissions[8]= (cur_stat.st_mode & S_IXOTH) ? 'x' : '-';


    
    struct passwd *p;
    struct group  *g;
    p = getpwuid(cur_stat.st_uid);
    g=getgrgid(cur_stat.st_gid);
    char *username = p->pw_name;
    char *groupname= g->gr_name;

    long long int file_sz= cur_stat.st_size;
    long long int no_of_links= cur_stat.st_nlink;



    char *tm_as_req = time_as_req(cur_stat.st_mtime);
 
    char *tmp =malloc(200);  
    sprintf(tmp,"%c%s %lld %s %s %lld %s %s\n",type,permissions,no_of_links,username,groupname,file_sz,tm_as_req,cur_dir->d_name);
    lines[i]= tmp;
    }
    printf("total %d\n",roundNo(((float)total/(float)2)));
    for(int p=0;p<cnt;p++)
     printf("%s",lines[p]);

    



}

int is_a_filter(const struct dirent *p)
{   if((p->d_name)[0] == '.')
    return 0;
    else 
    return 1;
}



void ls_implt(int cnt_pieces,char*pieces[])
{    
     char *cur_path;
     bool is_a=false,is_l=false;
     
     char *dirs[25];int no_of_dirs =0;
   
    for(int i=1;i<cnt_pieces;i++)
    {    if(pieces[i][0]=='-')
         {  for(int j=1;j<strlen(pieces[i]);j++)
            {  if(pieces[i][j]=='l')
                is_l=true;
               if(pieces[i][j]=='a')
                is_a=true;

            }
         }
       
         else
         { if(strlen(pieces[i])==1 && pieces[i][0]=='.')
           dirs[no_of_dirs]=cwd_name;
           else if(strlen(pieces[i])==2 && (strcmp(pieces[i],"..")==0))
            dirs[no_of_dirs]=detecting_parent(cwd_name);
             
           else
           dirs[no_of_dirs]=pieces[i];
           no_of_dirs++;  
         }
         
    }
    if(no_of_dirs ==0)
    {  dirs[no_of_dirs] = cwd_name;
       no_of_dirs++;

    }
   

     for (int i=0;i<no_of_dirs;i++)
    {   struct dirent **dir_list;
        int no_of_files;
        if(is_a == true && is_l ==true)
        {no_of_files=scandir(path(cwd_name,dirs[i]),&dir_list,NULL,&lexi_sort);
         display_long(dir_list,no_of_files,path(cwd_name,dirs[i]));}
        
        else if(is_a == true && is_l == false)
        {no_of_files=scandir(path(cwd_name,dirs[i]),&dir_list,NULL,&lexi_sort);
         display_short(dir_list,no_of_files,path(cwd_name,dirs[i]));}
        
        else if(is_a==false && is_l == true)
        {no_of_files=scandir(path(cwd_name,dirs[i]),&dir_list,&is_a_filter,&lexi_sort);
         display_long(dir_list,no_of_files,path(cwd_name,dirs[i]));}
        
        else if(is_a==false,is_l==false)
        {no_of_files=scandir(path(cwd_name,dirs[i]),&dir_list,&is_a_filter,&lexi_sort);
         display_short(dir_list,no_of_files,path(cwd_name,dirs[i]));}
      
       if(no_of_files == -1)
       {char * tmp = malloc(120* sizeof(char));
        strcpy(tmp,"Error Reg ");
        strcat(tmp,dirs[i]);
        perror(tmp);
       }
    }



}

main.c
#include "functions.h"
#include "headers.h"
int shell_pid;
int bgpid = 0;
bool first = true;

int cnt_pieces;
char *pieces[30];

void handler()
{    int corpse;
     int status;
     char *pr_name; 
   
     corpse = waitpid((pid_t)-1, &status, WNOHANG | WUNTRACED);
     
     
    if(corpse > 0)
    {  int i;
        for(i=0;i<cnt_of_bg;i++)
        if(bgarr[i].pid == corpse)
          break;
      if (WIFEXITED(status)) 
         {  int a = WEXITSTATUS(status); 
            if(a == 0)
              printf("\r%s with pid %d exited \n", bgarr[i].name ,corpse);
             else
               fprintf(stderr,"\r%s with pid %d exited with status %d \n", bgarr[i].name ,corpse,a);
               bgarr[i].is_active = 0;
           
            prompt();
             fflush(stdout);
         }
     
      else if (WIFSIGNALED(status)) 
        { //fprintf(stderr,"\n%s with pid %d killed by signal %d\n",bgarr[i].name ,corpse, WTERMSIG(status));
           //bgarr[i].is_active = 0;
           } 
     
      else if (WIFSTOPPED(status)) 
        fprintf(stderr,"\n%s with pid %d stopped by signal %d\n", bgarr[i].name ,corpse,WSTOPSIG(status));
     
      else if (WIFCONTINUED(status)) 
        fprintf(stderr,"\n%s with pid %d continued\n", bgarr[i].name ,corpse);
        
    }

}

void ctrlz_handler()
{   if(getpid()!=shell_pid)
    return;

  if(foregrndpid!=0)
  {  if(kill(foregrndpid,SIGTTIN)<0)
      {  perror("Error in kill");
          return;
      }
    if(kill(foregrndpid,SIGTSTP)<0)  
       { perror("Error in kill");  
         return ;
       }
        bgarr[cnt_of_bg].pid = foregrndpid;
        strcpy(bgarr[cnt_of_bg].name,pieces[0]);
        bgarr[cnt_of_bg].is_active =1; 
        cnt_of_bg++;
   }
  signal(SIGTSTP,ctrlz_handler);
}
void ctrlc_handler()
{  if(getpid()!=shell_pid)
  return; 
  if(foregrndpid!=0)
 {  if(kill(foregrndpid,SIGINT)<0)
        perror("Error in kill");
    foregrndpid=0;    
  fprintf(stderr,"\r");
 }

signal(SIGINT, ctrlc_handler);

}

void check_redir(int * cnt_pieces,char * pieces[])
{    int index_of_in = -1;
    int index_of_out = -1;
    int index_of_append =-1;
    int c = *cnt_pieces;
    for(int i=0;i<c;i++)
    { if(strcmp(pieces[i],"<")==0) 
        index_of_in=i;
      if(strcmp(pieces[i],">")==0)
      index_of_out=i;
      else if(strcmp(pieces[i],">>")==0)
      index_of_append =i;

    }
   if(index_of_out >=0)
   {  int open_fd = open(pieces[index_of_out+1], O_WRONLY | O_TRUNC | O_CREAT, 0644);
        if(open_fd < 0)
        {
            perror("Error reg file opening:");
            return;
        } 
        pieces[index_of_out]=NULL;pieces[index_of_out+1]=NULL;
        dup2(open_fd, 1);
        close(open_fd);
      c-=2; 
  }

    if(index_of_append >=0)
   {  int open_fd = open(pieces[index_of_append+1], O_WRONLY| O_APPEND | O_CREAT, 0644);
        if(open_fd < 0)
        {
            perror("Error reg file opening:");
            return;
        } 
        pieces[index_of_append]=NULL;pieces[index_of_append+1]=NULL;
        dup2(open_fd, 1);
        close(open_fd);
        c-=2;
    }
    

    if(index_of_in >=0)
   {  int open_fd = open(pieces[index_of_in+1], O_RDONLY);
        if(open_fd < 0)
        {
            perror("Error reg file opening:");
            return;
        } 
        pieces[index_of_in]=NULL;pieces[index_of_in+1]=NULL;
        dup2(open_fd, 0);
        close(open_fd);
     c-=2;
  }

 *cnt_pieces = c;

}




int fn(char *p)
{   
  //int actual_in = dup(0);
  int actual_out = dup(1);
 
  separating_str(p,pieces,&cnt_pieces);

  check_redir(&cnt_pieces,pieces);
 

  
  if(strcmp(pieces[0],"quit")==0)
  return 0; 
  if(strcmp(pieces[cnt_pieces-1],"&")==0)
    { pieces[cnt_pieces-1] = NULL;
         cnt_pieces--;
         background =1;
    }
  else
      background = 0;
     
      
  if(background && is_builtin_cmd(cnt_pieces,pieces))
       background = 0;
       
  if(background)
  { pid_t pid=fork();
    if(pid==0)
    { setpgid(0,0);
      execute_theline_other(cnt_pieces,pieces);
    }
    
      printf("started %s with pid %d\n",pieces[0],pid);
      bgarr[cnt_of_bg].pid = pid;
      strcpy(bgarr[cnt_of_bg].name,pieces[0]);
      bgarr[cnt_of_bg].is_active =1;
      
     cnt_of_bg++;  
      
  }

  else if (!background)
      {  if(is_builtin_cmd(cnt_pieces,pieces))
          execute_theline_buitin(cnt_pieces,pieces);
         else 
        {  int k =fork();
          if(k<0)
          { perror("Error in forking:");
            return 0;
          }
      
          if(k==0)
          { 
            execute_theline_other(cnt_pieces,pieces);
          
          }
          if(k > 0)
          { foregrndpid = k;
            
            
           
            waitpid(-1,NULL,WUNTRACED);

           /*if(WIFSTOPPED(ret_status))
           {  int ret_status;
           }*/
            foregrndpid =0;
           
          } 
         }
      }
  
  
  
   dup2(actual_out,1);
//   dup2(actual_in,0);   
return 1;
}



int main()
{  
    size_t buff_sz = 1000;
    buffer= (char *)malloc(buff_sz*sizeof(char));
    buffer_dup=(char *)malloc(buff_sz*sizeof(char));
    intialize();
    
    shell_pid = getpid();
    while (1)
    {     free(buffer);
          free(buffer_dup);

         signal(SIGCHLD,handler);
          signal(SIGTSTP,ctrlz_handler);
          signal(SIGINT,ctrlc_handler);
          //signal(EOF,ctrld_handler);
        sub_intialize();
       // printf("pid is %d |",getpid());
       buffer= (char *)malloc(buff_sz*sizeof(char));
    buffer_dup=(char *)malloc(buff_sz*sizeof(char));
        prompt();
       int b = getline(&buffer,&buff_sz,stdin);
       if(b<=0)
       { printf("\nShell is been exited\n");
         return 0;
       }
        store_history();
        strcpy(buffer_dup,buffer);
      
      char *parts[80],*pieces_pipes[20];
      int cnt_no_commands,cnt_based_on_pipes;

     

      // based on ;
      separating_buff(buffer_dup,parts,&cnt_no_commands);
      int a=1;
      for(int i = 0;i<cnt_no_commands;i++)
      {  
        separating_based_on_pipes(parts[i],pieces_pipes,&cnt_based_on_pipes);

        int no_of_pipes = cnt_based_on_pipes - 1;
        int actual_in= dup(0);
        int actual_out = dup(1);

        int j=0;
        for(j=0;j< no_of_pipes;j++)
        { 
          int fd_pipes[2];
          pipe(fd_pipes);
          
          int p =fork();
          if(p == 0)
          { dup2(fd_pipes[1],1);
            close(fd_pipes[0]);
            close(fd_pipes[1]);
           
            fn(pieces_pipes[j]);
             abort();           
          } 
          else if (p >0)
          { wait(NULL);
            dup2(fd_pipes[0],0);
            close(fd_pipes[0]);
            close(fd_pipes[1]);
          }
     

        }
        if(j == no_of_pipes)
        {  dup2(actual_out,1);
           a =fn(pieces_pipes[j]);
          if(!a)
          break;
        }
       

        
       dup2(actual_in,0);
       dup2(actual_out,1); 
      
   
     strcpy(lwd,cwd_name);
     }
    if(a==0)
     {  printf("Shell is been exited\n");
       break;}
    }
return 0;
}


history.c
#include "functions.h"
#include "headers.h"



void store_history()
{  
    FILE * frmfptr;
    FILE * tofptr;

   int present_line_no,tobereplaced_no;
   char path[200];
   strcpy(path,base_dir);
   strcat(path,"/history.txt");
      
   int buf_sz =1000;

   char tmp_buf[buf_sz];
   char inpt_line[buf_sz];
   
   strcpy(inpt_line,buffer);

   frmfptr = fopen(path,"r");
   if(frmfptr == NULL)
   {frmfptr = fopen("history.txt","w+");
    char *tmp = "0\n";
    fputs(tmp,frmfptr);
    fseek(frmfptr,0,SEEK_SET);
   }
   tofptr = fopen("replace.tmp","w");
   //fseek(frmfptr,0,SEEK_SET);
  
  char *num_buff = malloc(5);
  index_of_hist = 0; 
  present_line_no = 0 ;

  
  fgets(tmp_buf,buf_sz,frmfptr);

 
  //printf("index as read from file %s,%ld\n",tmp_buf,strlen(tmp_buf));

  for(int i=0;i<strlen(tmp_buf)-1;i++)
  { index_of_hist = index_of_hist*10+(tmp_buf[i]-'0');
  }


  if(index_of_hist == 20)
   index_of_hist = 0;
  index_of_hist++;
  
  //printf("index of history %da\n",index_of_hist);
  sprintf(num_buff,"%d\n",index_of_hist);
  fputs(num_buff,tofptr);
   
 present_line_no =0;
 
  while(fgets(tmp_buf,buf_sz,frmfptr)!=NULL)
  {  present_line_no++;
     
     if(present_line_no == index_of_hist)
     fputs(inpt_line,tofptr);
     else
      fputs(tmp_buf,tofptr);
  }
  if(present_line_no !=20)
  {
    fputs(inpt_line,tofptr);
    
  } 
  //if(index_of_hist==20)
  //is_historyfull=1;
  fclose(frmfptr);
  fclose(tofptr);

   remove(path);
   rename("replace.tmp",path);
   

}


   

void hist_implt(int cnt_pieces, char *pieces[])
{ 
    FILE * frmfptr;
   char path[200];
   strcpy(path,base_dir);
   strcat(path,"/history.txt");
   frmfptr = fopen(path,"r");

    int c=0 ,index =0 ;
    if(cnt_pieces == 1)
    c =10;
    else 
    {   for(int j=0;j<strlen(pieces[1]);j++)
          c = c*10 + (pieces[1][j]-'0');
    }
    char *l_of_history[21];

   int k;
   for(k=0;k<21;k++)
   {  l_of_history[k]=malloc(1000);
      
      if(fgets(l_of_history[k],1000,frmfptr)==NULL)
          break;
   
   }
   if(k>=20)
   is_historyfull = 1;


  for(int i=0;i<strlen(l_of_history[0])-1;i++)
  { index = index*10+(l_of_history[0][i]-'0');
  }
  
  int sp = index - c;
  //printf("%d %d\n",sp,is_historyfull);
  if(sp < 0 )
  {  if(is_historyfull)
     {  sp = 20 +sp;
      
      for(int k =sp+1;k<=20;k++)
      { printf("%s",l_of_history[k]);
        
        if(k == 20)
        {  k=1; break;}
      }
      for(k =1;k<=index;k++)
      { printf("%s",l_of_history[k]);
        
        
      }
     
     }
     else
     {   sp=0;
       for(int k =sp+1;k<=index;k++)
       printf("%s",l_of_history[k]);
        
     }
  }
  else
  {  for(int k =sp+1;k<=index;k++)
       printf("%s",l_of_history[k]);
      
  }
  
fclose(frmfptr);


}

     






   













functions.h
#ifndef __FUNCTIONS_H
#define __FUNCTIONS_H
void handler();
void intialize();
void sub_intialize();
void prompt();
void separating_buff(char * buff, char *parts[],int * num);
void separating_based_on_pipes(char * buff, char *pieces_pipes[],int *num);
void separating_str(char * buff, char *pieces[],int * num);

void execute_theline_buitin(int cnt_pieces, char *pieces[]);
void execute_theline_other(int cnt_pieces, char *pieces[]);
void cd_implt(int cnt_pieces, char *pieces[]);
void echo_implt(int cnt_pieces, char *pieces[]);
void ls_implt(int cnt_pieces, char *pieces[]);
void pinfo_implt(int cnt_pieces, char *pieces[]);
int is_builtin_cmd(int cnt_pieces, char *pieces[]);
void store_history();
void hist_implt(int cnt_pieces, char *pieces[]);
void setenv_implt(int cnt_pieces, char *pieces[]);
void unsetenv_implt(int cnt_pieces, char *pieces[]);
void getenv_implt(int cnt_pieces, char *pieces[]);
void jobs_implt(int cnt_pieces, char *pieces[]);
void kjob_implt(int cnt_pieces, char *pieces[]);
void fg_implt(int cnt_pieces,char *pieces[]);
void bg_implt(int cnt_pieces,char *pieces[]);
void overkill_implt(int cnt_pieces, char *pieces[]);
int fn(char *p);


int background;
int bgprpidarr[1024][1];
char *bgprpidnamearr[1024];
int cnt_of_bg;
int index_of_hist;
int is_historyfull;
int foregrndpid;

struct passwd *fas;
char base_dir[200];
char cwd_name[200];
char *cwd_asreq;
char hstnam[50];
char *buffer;
char *buffer_dup;
char lwd[200];

typedef struct bgp
{  int pid;
   int is_active;
   char name[100];
   char status[100];
}bgp;

bgp bgarr[1024];


#endif

functions.c
#include "functions.h"
#include "headers.h"


int convert_to_int(char * str)
{    int t=0,i=0,s=1;
     if(str[i]=='-')
     { s=-1;i++;}

     for(;i < strlen(str);i++)
   {   t = t*10 + (str[i]-'0');
   }
return t*s;

}

int is_builtin_cmd(int cnt_pieces, char *pieces[])
{  if(strcmp(pieces[0],"cd")==0 || strcmp(pieces[0],"echo")==0 || strcmp(pieces[0],"ls")==0 || strcmp(pieces[0],"pinfo")==0 
    ||strcmp(pieces[0],"history")==0 || strcmp(pieces[0],"setenv")==0 || strcmp(pieces[0],"unsetenv")==0 || strcmp(pieces[0],"getenv")==0
    || strcmp(pieces[0],"jobs")==0 || strcmp(pieces[0],"kjob")==0 || strcmp(pieces[0],"fg")==0 || strcmp(pieces[0],"bg")==0 || strcmp(pieces[0],"overkill")==0)
    return 1;
    else
    return 0;
}




void intialize()
{    
    gethostname(hstnam,50);
    fas= getpwuid(getuid());
    getcwd(base_dir,sizeof(base_dir));
    cnt_of_bg = 0;
    index_of_hist =0;
    foregrndpid =0 ;

}

void sub_intialize()
{   getcwd(cwd_name,sizeof(cwd_name));
    if(strncmp(cwd_name,base_dir,strlen(base_dir))==0)
     cwd_asreq = &cwd_name[strlen(base_dir)];
    else
     cwd_asreq = cwd_name;
   
}

void prompt() 
{    
  printf("<%s@%s:~%s>",fas->pw_name,hstnam,cwd_asreq);

}



void separating_buff(char * buff, char *parts[],int *num)
{   
   int c=0;
   char * token = strtok(buff,";\n");

   while( token!=NULL )
   {
       parts[c] = strdup(token);
       c++;
       token = strtok(NULL,";\n");
   
   }
   parts[c]=NULL;   
   *num = c;
}

void separating_based_on_pipes(char * buff, char *pieces_pipes[],int *num)
{   
   int c=0;
   char * token = strtok(buff,"|");

   while( token!=NULL )
   {
       pieces_pipes[c] = strdup(token);
       c++;
       token = strtok(NULL,"|");
   
   }
   pieces_pipes[c]=NULL;   
   *num = c;
}

void separating_str(char * buff, char *pieces[],int *num )
{  int c =0;
   char * token = strtok(buff," \t\n"); 

   while( token!=NULL )
   {
       pieces[c] = strdup(token);
       c++;
       token = strtok(NULL," \t\n");
   
   }
    pieces[c]=NULL;
   *num = c;
   
}


void execute_theline_other(int cnt_pieces, char *pieces[])
{       int k =0; 
        k=execvp(pieces[0],pieces);
        if(k==-1)
        printf("%s: command not found\n",pieces[0]);
       // perror("error:");
        exit(0); 
       

}



void execute_theline_buitin(int cnt_pieces, char *pieces[])
{     if(strcmp(pieces[0],"cd")==0)
        cd_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"echo")==0)
           echo_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"ls")==0)
          ls_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"pinfo")==0)
          pinfo_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"history")==0)
          hist_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"setenv")==0)
           setenv_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"unsetenv")==0)
          unsetenv_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"getenv")==0)
          getenv_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"jobs")==0)
          jobs_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"kjob")==0)
          kjob_implt(cnt_pieces,pieces);
       else if(strcmp(pieces[0],"fg")==0)
          fg_implt(cnt_pieces,pieces);
        else if(strcmp(pieces[0],"bg")==0)
          bg_implt(cnt_pieces,pieces);   
      else if(strcmp(pieces[0],"overkill")==0)
          overkill_implt(cnt_pieces,pieces);   

}


void echo_implt(int cnt_pieces, char *pieces[])
{   for(int i=1;i<cnt_pieces;i++)
    printf("%s ",pieces[i]);
    printf("\n");
    
}

void pwd_implt(int cnt_pieces, char *pieces[])
{   printf("%s",cwd_name);

}

void cd_implt(int cnt_pieces, char *pieces[])
{    if(cnt_pieces>1)
    {  if(strcmp(pieces[1],"~") == 0) 
        chdir(base_dir);
        
       else if(strcmp(pieces[1],".")==0)
        {}
       else if(strcmp(pieces[1],"-")==0)
       {  printf("%s\n",lwd);
          chdir(lwd);
       }
       else
       { if(chdir(pieces[1])==-1)
           perror("Error");
           
       }
    }

    else
      {  
          chdir(base_dir); 
       
      }
      
  
}



void pinfo_implt(int cnt_pieces, char *pieces[])
{   int pid;
    

     if(cnt_pieces > 1)
      {   pid = convert_to_int(pieces[1]);
          
      }
          
     else
      { pid=getpid();}
   
   char filename_stat[200];
  sprintf(filename_stat,"/proc/%d/stat",pid);

  //  printf("%s",filename_stat);
   FILE *f = fopen(filename_stat,"r");
    if( f == NULL )
    perror("Error:");
   
   char state[2];
   long unsigned int vm_sz;
   int t = fscanf(f,"%*d%*s%s%*d%*d%*d%*d%*d%*u%*u%*u%*u%*u%*u%*u%*d%*d%*d%*d%*d%*d%*u%lu",state,&vm_sz);


   fclose(f);

   char filename_exe[200];
   sprintf(filename_exe,"/proc/%d/exe",pid);
   char *exe_path= malloc(500);
   int k= readlink(filename_exe,exe_path,500);
    if(k ==-1)
   perror("Error:");
   
   exe_path[k]='\0';
  
   printf("pid -- %d\nProcess Status -- %s\nmemory -- %lu\nExecutable Path -- %s\n",pid,state,vm_sz,exe_path);



}

void setenv_implt(int cnt_pieces, char *pieces[])
{  if(cnt_pieces==3) 
   { int tmp = setenv(pieces[1],pieces[2],1);
     if(tmp < 0)
     perror("Error reg setenv");
   }
   else if(cnt_pieces==2)
   { int tmp = setenv(pieces[1],pieces[2],1);
     if(tmp < 0)
     perror("Error reg setenv");
 
   }
   else
   {
      printf("No of arguments needs to be 2 for setenv\n");
   }
   
}
void unsetenv_implt(int cnt_pieces, char *pieces[])
{ if(cnt_pieces==2)
  {  int temp =unsetenv(pieces[1]);
     if(temp < 0)
     perror("Error reg unsetenv");
     }
  else
  {
     printf("No of arguments needs to be 1 for unsetenv");
  }
  
}
void getenv_implt(int cnt_pieces, char *pieces[])
{  char * t = getenv(pieces[1]);
   if(t == NULL)
   printf("environment variable not found\n");
   else 
   printf("%s\n",t);
}


void jobs_implt(int cnt_pieces, char *pieces[])
{  int j; 
   for(int i=0,j=0;i<cnt_of_bg;i++)
     {  if(bgarr[i].is_active == 1)
       {  j++;
          char filename_stat[200];
       //  printf("The pid is %d\n",bgarr[i].pid);
          sprintf(filename_stat,"/proc/%d/stat",bgarr[i].pid);
          
         FILE *f = fopen(filename_stat,"r");
    
         if( f == NULL )
        { perror("Error:");
         return;
         }
          char state[2];
         int t = fscanf(f,"%*d%*s%s",state);
         
        fclose(f);
       
       if(t<=0)
        { printf("error in status checking\n");
          return ;
        }
       
       
       strcpy(bgarr[i].status,"Running");
       if(state[0]=='S')
       strcpy(bgarr[i].status,"Running");
       else if(state[0]=='T')
       strcpy(bgarr[i].status,"Stopped");       
        
       
       printf("[%d] %s %s [%d]\n",j,bgarr[i].status,bgarr[i].name,bgarr[i].pid);
      
       }

     }



}
void kjob_implt(int cnt_pieces, char *pieces[])
{   int j_num = 0; int sig=0;
   
   for(int i=0;i<strlen(pieces[1]);i++)
    j_num = j_num*10+(pieces[1][i]-'0');
   
   for(int i=0;i<strlen(pieces[2]);i++)
    sig = sig*10+(pieces[2][i]-'0');
   
   int j;
   for(int i=0,j=0;i<cnt_of_bg;i++)
   { if(bgarr[i].is_active == 1 )
     { j++;
       if(j == j_num)
      { int a = kill(bgarr[i].pid,sig);
        bgarr[i].is_active=0;
       if(a == -1)
        perror("Reg kjob");
        break;
      }
     }
   
   }
  // printf("%d %d\n",j_num,sig);


}

void fg_implt(int cnt_pieces, char*pieces[])
{   int j_num =0;
  
  for(int i=0;i<strlen(pieces[1]);i++)
    j_num = j_num*10+(pieces[1][i]-'0');

   int j=0, pid_of_req=-5;
  int k;
  for(k=0;k<cnt_of_bg;k++)
  {  if(bgarr[k].is_active == 1)
     { j++;
       if(j == j_num)
       {
         pid_of_req = bgarr[k].pid;
         break;
       }

     }
   }
   if(pid_of_req == -5)
   {  printf("No job found with the given job num\n");
      return;
   }
   signal(SIGCHLD, SIG_IGN);
    
   int temp = kill(pid_of_req,SIGCONT);
   if(temp<0)
   {  perror("Error reg kill");
   }
   foregrndpid = pid_of_req;
   bgarr[k].is_active=0;

   int ret_status;
   waitpid(-1,&ret_status,WUNTRACED);
   
   if(WIFSTOPPED(ret_status))
   {   bgarr[k].is_active=1;

   }
   signal(SIGCHLD, handler);

}

void bg_implt(int cnt_pieces,char *pieces[])
{  int j_num =0;
  
  for(int i=0;i<strlen(pieces[1]);i++)
    j_num = j_num*10+(pieces[1][i]-'0');

   int j=0, pid_of_req=-5;
  int k;
  for(k=0;k<cnt_of_bg;k++)
  {  if(bgarr[k].is_active == 1)
     { j++;
       if(j == j_num)
       {
         pid_of_req = bgarr[k].pid;
         break;
       }


     }
   }
   if(pid_of_req == -5)
   {  printf("No job found with the given job num\n");
      return;
   }
   int temp= kill(pid_of_req,SIGCONT);
    if(temp<0)
    perror("Error reg kill:");
}

void overkill_implt(int cnt_pieces, char *pieces[])
{  
   for(int i=0;i<cnt_of_bg;i++)
   {   if(bgarr[i].is_active == 1)
     {  if( kill(bgarr[i].pid,SIGKILL) <0)
         perror("Error in overkill");
        bgarr[i].is_active=0;
     }
   }


}